### DISCUSSION TITLE
decode function only works when connected to Thonny?

### DISCUSSION THREAD

**Discussion by sanamon** [micropython/micropython#11073] [RP2040 / Pico] [answered]
Hi,

I have this code that works fine when Thonny is connected and run from within Thonny:
```
def waitResp(timeout=3000):
    prvMills = utime.ticks_ms()
    resp = b""
    while (utime.ticks_ms()-prvMills)<timeout:
        if uart.any():
            resp = b"".join([resp, uart.read(1)])
    print((resp).decode())
```

However if run by just applying power to the RPi Pico it hangs unless the ".decode()" is removed from the last line. But that of course has other implications... 
The snippet above is taken from the modem example code, but slightly adapted:  [https://www.waveshare.com/wiki/File:Pico-SIM7020X-NB-loT-Demo-Code.zip]

The Pico talks to a SIM7020E IoT-NB modem via the uart.

MicroPython v1.19.1-966-g05bb26010 on 2023-03-13; Raspberry Pi Pico W with RP2040

Any ideas?

Best Regards
Jonas

**Comment by robert-hh** [comment, marked as answer]
> However if run by just applying power to the RPi Pico it hangs unless the ".decode()" is removed from the last line. 

In that case use a simple terminal emulator like mpremote or putty. Then you will see the error message.

**Reply by sanamon** [reply]
Great idea, thanks! 

The error is as follows:
```
Traceback (most recent call last):
  File "main.py", line 134, in <module>
  File "main.py", line 72, in waitResp
UnicodeError: 
MicroPython v1.19.1-966-g05bb26010 on 2023-03-13; Raspberry Pi Pico W with RP2040
```

I don't know if **UnicodeError:**  really helps though.  **Perhaps this is a micropython issue?** 

I found a workaround: use str() to convert the bytes to a string instead.

**Reply by robert-hh** [reply]
Unicode is a specific character set. decode() requires it's argument to be unicode characters. An arbitrary byte sequence will not always consist of unicode characters. That's nothing specific to MicroPython, and using Thonny should not make a difference.
P.S.: The way to accumulate the resp can be made simpler:
```
        if uart.any():
            resp += uart.read()
```

**Reply by sanamon** [reply]
You are correct. Strangely it does make a difference if being run through Thonny or not. Thanks again.
